bundles/org.openhab.binding.shelly/src/main/resources/ESH-INF/thing/protected.xml

```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="shelly"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="shelly-protected">
```

Comment:
This feature is was well received by the community. When the device is password protected you can't access the settings even not the device type AND mode. Therefore I create this "placeholder" so the user has the chance to set user+password and run the discovery again.
The device type is derived from the mDNS service name, but I can't get the mode. This is required, because devices could have different modes resulting in different channel layouts. e.g. the Shelly 2 has a Relay and a Roller mode. I know that it is a shelly 2, but I can't assign the proper thing type, because I can't access the mode information.

Again, this functionality was well received by the users and should not be removed. In fact OH 2.x is missing some kind of user notification so the binding could send this information  to the UI while the user is running the discovery.

Reply:
Sorry, but you must not use the Inbox as a notification system.
Every entry in the inbox must correspond to some real device that has been discovered.
So if you find 5 password protected shelly devices, you should add 5 inbox entries providing their names and maybe of an "unidentified shelly device" Thing type. If those get approved, the thing is created and it should be OFFLINE->CONFIGURATION ERROR with a message that the credential parameters need to be set. Once the user does this, the magic method that you need is [changeThingType](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L593) - once you know, what exact device it is, this method will automatically add all the channels of that type to the existing thing.
Btw, you will have to have such a mechanism anyhow for all devices that do not use the default username/password, right?

Reply:
> Sorry, but you must not use the Inbox as a notification system.
> Every entry in the inbox must correspond to some real device that has been discovered.
> So if you find 5 password protected shelly devices, you should add 5 inbox entries providing their names and maybe of an "unidentified shelly device" Thing type. If those get approved, the thing is created and it should be OFFLINE->CONFIGURATION ERROR with a message that the credential parameters need to be set. Once the user does this, the magic method that you need is [changeThingType](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L593) - once you know, what exact device it is, this method will automatically add all the channels of that type to the existing thing.
> Btw, you will have to have such a mechanism anyhow for all devices that do not use the default username/password, right?

There is no need for. When the device is not password protected I could access device type and mode, the binding selects the corresponding thing type and the thing is created in the Inbox.

> So if you find 5 password protected shelly devices, you should add 5 inbox entries providing their names and maybe of an "unidentified shelly device" Thing type. 

?? That's what I'm doing: Creating a Thing of type unidentified

> If those get approved, the thing is created and it should be OFFLINE->CONFIGURATION ERROR with a message that the credential parameters need to be set. Once the user does this, the magic method that you need is [changeThingType](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L593) - once you know, what exact device it is, this method will automatically add all the channels of that type to the existing thing.

That sounds complex. Is there a binding working with a similar logic?

> Btw, you will have to have such a mechanism anyhow for all devices that do not use the default username/password, right?
In this case the device accepts the credentials and it "feels" like a normal thing initialization.

Reply:
> There is no need for. When the device is not password protected

I meant, when the device IS password protected but NOT using the default password.

> That's what I'm doing: Creating a Thing of type unidentified

Ok, that's the right start then. But those should not be seen as a placeholder, but as the device itself - it must be possible that the user approves those entries and works with them.

> Is that correct?

Yes, absolutely. 

> That sounds a bit complex.

Not really, because you'll need this logic anyhow. Think of a device where the user might have changed the password: Your existing thing should notice that it cannot authenticate anymore and set the thing status to CONFIGURATION_ERROR. And upon changed credentials in the configuration you will also always need to reinitialize the Thing with the new credentials.

> Is there a binding working with a similar logic?

[A few](https://github.com/openhab/openhab2-addons/search?q=changeThingType&unscoped_q=changeThingType) - plus the Z-Wave binding uses this feature extensively.

Reply:
ok, more or less done. Needs testing

Reply:
ok, I have it running in general
- Thing discovered, will be placed with an "shellyunknown" thing type in in Inbox
- User acceots it and thing gets active
- accessing API will fail, thing goes OFFLINE with COMMUNICATION_ERROR ...
- User edits uid+pw, access to API now successful
- Thing type will be changed with the call mentioned, e.g. shelly2-relay
Now the strange part: Thing gets ONLINE, channels are there and working
BUT the shingtype is still "shellyunknown" with the Description of that thing type

- How to force an update of that information according to the info included in the regular thing type definition?
- I flagged the userId and password in the parameters with required="true". However, the thing gets accepted without asking for those fields. I made this also for the Rachio binding and then PaperUI after ticking the checkmark in the Inbox brings me to the Thing configuration before finally creating thething. Now I need to accept it in the Inbox, need to go to Confgurarion:Things and edit it there.

@Kai: Any idea on those topics?

---

...
	<thing-type id="shellyunknown">
		<label>Password protected Shelly device</label>
		<description>A password protected device was detected. Add Thing and set credentials.</description>
		<config-description-ref uri="thing-type:shelly:unknown" />
	</thing-type>
</thing:thing-descriptions>

	<config-description uri="thing-type:shelly:unknown">
		<parameter name="userId" type="text" required="true">
			<label>UserID</label>
			<description>User ID for API access.</description>
		</parameter>
		<parameter name="password" type="text" required="true">
			<label>Password</label>
			<description>Password for API access.</description>
			<context>password</context>
		</parameter>
		<parameter name="deviceIp" type="text" required="true">
			<label>Device IP Address</label>
			<description>IP-Address of the Shelly device.</description>
			<context>network-address</context>
		</parameter>
	</config-description>

Reply:
Great, sounds like a wonderful progress already 👍.

> BUT the shingtype is still "shellyunknown" with the Description of that thing type

Sounds strange indeed. Can you post a screenshot and post the output of `smarthome:things show <thingUID>`?

> I flagged the userId and password in the parameters with required="true".

I have seen this effect as well somewhere. I'd consider it a bug in the Paper UI, so best ignore it for now.